File: python/test/test_model_validator.py
# -*- coding: utf-8 -*-
"""
ModelValidator — 独立模型验证

无需搭建 Pipeline，即可验证模型的：
  1. 加载与元信息检查 (tensor shapes, dtypes)
  2. 原始张量推理 (raw tensor in/out, 无前后处理)
  3. 端到端推理 (image -> preproc -> infer -> postproc -> detections)
  4. 性能基准测试 (latency, p99, fps)

用法:
    python test_model_validator.py \
        --model /path/to/yolov8s.engine \
        --device cuda \
        --image /path/to/test.jpg \
        --postproc-config /path/to/yolo_coco.json

    # 仅检查模型信息 (不推理)
    python test_model_validator.py --model /path/to/model.engine --info-only

    # 使用合成图像 (无需真实图片)
    python test_model_validator.py --model /path/to/model.engine --device cuda
"""

import argparse
import os
import sys
import time

import numpy as np
import cv2

try:
    import vstream
except ImportError:
    print("[ERROR] Cannot import vstream. Build the Python module first:")
    print("  ./build.sh --python")
    sys.exit(1)


def create_synthetic_image(width: int = 1280, height: int = 720) -> np.ndarray:
    """Create a synthetic BGR test image with some shapes."""
    img = np.zeros((height, width, 3), dtype=np.uint8)
    img[:] = (50, 100, 150)  # BGR background
    cv2.rectangle(img, (100, 100), (300, 300), (0, 255, 0), 3)
    cv2.rectangle(img, (500, 200), (800, 500), (0, 0, 255), -1)
    cv2.circle(img, (1000, 400), 80, (255, 0, 0), -1)
    return img


def print_separator(title: str):
    print("\n" + "=" * 60)
    print(f"  {title}")
    print("=" * 60)


#  Test 1: Load model & inspect metadata
def test_load_and_info(validator: vstream.ModelValidator):
    print_separator("Step 1: Load Model & Inspect Metadata")

    if not validator.load():
        print("[FAIL] Model load failed.")
        return False

    assert validator.is_loaded(), "is_loaded() should be True after load()"
    print("[OK] Model loaded successfully.")

    info = validator.get_model_info()
    print(f"  Model path : {info.model_path}")
    print(f"  Device     : {info.device_type} (id={info.device_id})")
    print(f"  Batch size : {info.batch_size}")
    print(f"  Input shape: {info.width}x{info.height}x{info.channel}")
    print(f"  Inputs     : {len(info.inputs)}")
    for i, t in enumerate(info.inputs):
        print(f"    [{i}] name={t.name}, shape={t.shape}, dtype={t.dtype}")
    print(f"  Outputs    : {len(info.outputs)}")
    for i, t in enumerate(info.outputs):
        print(f"    [{i}] name={t.name}, shape={t.shape}, dtype={t.dtype}")

    return True


#  Test 2: Raw tensor inference
def test_raw_infer(validator: vstream.ModelValidator):
    print_separator("Step 2: Raw Tensor Inference (no preproc/postproc)")

    info = validator.get_model_info()

    # Build random float32 inputs matching tensor shapes
    inputs = []
    for t in info.inputs:
        count = 1
        for d in t.shape:
            count *= d
        data = np.random.rand(count).astype(np.float32)
        inputs.append(data)
        print(f"  Input '{t.name}': shape={t.shape}, elements={count}")

    outputs = validator.infer(inputs)

    if not outputs:
        print("[FAIL] Inference returned empty results.")
        return

    print(f"  [OK] Got {len(outputs)} output tensor(s):")
    for i, out in enumerate(outputs):
        arr = np.asarray(out)
        print(f"    Output[{i}]: size={arr.size}, "
              f"min={arr.min():.4f}, max={arr.max():.4f}, "
              f"mean={arr.mean():.4f}")
        # Check for NaN/Inf
        if np.any(np.isnan(arr)):
            print(f"    [WARN] Output[{i}] contains NaN!")
        if np.any(np.isinf(arr)):
            print(f"    [WARN] Output[{i}] contains Inf!")


#  Test 3: End-to-end inference
def test_run_e2e(validator: vstream.ModelValidator, image: np.ndarray,
                 preproc_name: str, postproc_name: str,
                 postproc_config: str = ""):
    print_separator("Step 3: End-to-End (image -> preproc -> infer -> postproc)")

    postproc_params = {}
    if postproc_config:
        postproc_params["config_file"] = postproc_config

    print(f"  Image       : {image.shape[1]}x{image.shape[0]} (WxH)")
    print(f"  Preproc     : {preproc_name}")
    print(f"  Postproc    : {postproc_name}")
    if postproc_config:
        print(f"  Postproc cfg: {postproc_config}")

    result = validator.run_e2e(
        image, preproc_name, postproc_name,
        postproc_params=postproc_params
    )

    if result.error:
        print(f"  [ERROR] {result.error}")
        if not postproc_config:
            print("  (hint: provide --postproc-config for postproc label names)")
        return

    print(f"  [OK] {len(result.detections)} detections in {result.latency_ms:.2f} ms")
    for i, det in enumerate(result.detections[:20]):
        print(f"    det[{i}]: class={det.class_id}, name='{det.class_name}', "
              f"score={det.score:.4f}, bbox=[{det.x:.3f}, {det.y:.3f}, "
              f"{det.w:.3f}, {det.h:.3f}]")
    if len(result.detections) > 20:
        print(f"    ... and {len(result.detections) - 20} more")


#  Test 4: Benchmark
def test_benchmark(validator: vstream.ModelValidator, image: np.ndarray,
                   preproc_name: str, postproc_name: str,
                   postproc_config: str = "",
                   warmup: int = 10, runs: int = 50):
    print_separator(f"Step 4: Benchmark (warmup={warmup}, runs={runs})")

    postproc_params = {}
    if postproc_config:
        postproc_params["config_file"] = postproc_config

    results = validator.benchmark(
        image, preproc_name, postproc_name,
        postproc_params=postproc_params,
        warmup_runs=warmup,
        test_runs=runs,
        batch_sizes=[1]
    )

    for r in results:
        print(f"  Batch={r.batch_size}: "
              f"avg={r.avg_ms:.2f}ms, "
              f"min={r.min_ms:.2f}ms, "
              f"max={r.max_ms:.2f}ms, "
              f"p99={r.p99_ms:.2f}ms, "
              f"fps={r.fps:.1f}, "
              f"errors={r.error_count}")



#  Test 5: Error path tests (no model needed)
def test_error_paths():
    print_separator("Step 0: Error Path Tests (no model needed)")

    # Construct with non-existent model
    v = vstream.ModelValidator("/nonexistent/model.engine", "cpu", 0)
    assert not v.is_loaded(), "Should not be loaded before load()"
    print("  [OK] Constructed validator, not loaded")

    # Load should fail
    ok = v.load()
    assert not ok, "Load should fail for non-existent model"
    print("  [OK] Load() returned False for non-existent model")

    # GetModelInfo should return empty info
    info = v.get_model_info()
    assert info.batch_size == 0
    assert len(info.inputs) == 0
    print("  [OK] get_model_info() returns empty before load")

    # Infer should return empty
    outputs = v.infer([np.zeros(1, dtype=np.float32)])
    assert len(outputs) == 0
    print("  [OK] infer() returns empty before load")

    # RunE2E should return error
    dummy_img = np.zeros((640, 640, 3), dtype=np.uint8)
    result = v.run_e2e(dummy_img, "Pre_YOLO_CPU_v2", "Post_YOLOv8_CPU_v2")
    assert result.error, "run_e2e should have error before load"
    print(f"  [OK] run_e2e() returns error: '{result.error}'")


def main():
    parser = argparse.ArgumentParser(
        description="vStream ModelValidator — standalone model validation"
    )
    parser.add_argument("--model", type=str, default="",
                        help="Path to model file (.engine / .rknn)")
    parser.add_argument("--device", type=str, default="cuda",
                        choices=["cuda", "rockchip"],
                        help="Device type (default: cuda)")
    parser.add_argument("--device-id", type=int, default=0,
                        help="Device ID (default: 0)")
    parser.add_argument("--image", type=str, default="",
                        help="Path to test image (BGR). If empty, uses synthetic image.")
    parser.add_argument("--postproc-config", type=str, default="",
                        help="Path to postproc config JSON (e.g. yolo_coco.json)")
    parser.add_argument("--preproc", type=str, default="Pre_YOLO_CPU_v2",
                        help="Preproc class name (default: Pre_YOLO_CPU_v2)")
    parser.add_argument("--postproc", type=str, default="Post_YOLOv8_CPU_v2",
                        help="Postproc class name (default: Post_YOLOv8_CPU_v2)")
    parser.add_argument("--info-only", action="store_true",
                        help="Only load model and print info, skip inference")
    parser.add_argument("--no-benchmark", action="store_true",
                        help="Skip benchmark step")
    parser.add_argument("--warmup", type=int, default=10,
                        help="Benchmark warmup runs (default: 10)")
    parser.add_argument("--runs", type=int, default=50,
                        help="Benchmark test runs (default: 50)")
    args = parser.parse_args()

    # --- Error path tests (always run, no model needed) ---
    test_error_paths()

    if not args.model:
        print("\n[INFO] No --model provided. Exiting after error path tests.")
        print("[INFO] To test with a real model, run:")
        print("  python test_model_validator.py --model /path/to/model.engine --device cuda")
        return

    validator = vstream.ModelValidator(
        args.model, args.device, args.device_id, 0
    )

    # --- Step 1: Load & info ---
    if not test_load_and_info(validator):
        print("\n[FAIL] Cannot continue without a loaded model.")
        sys.exit(1)

    if args.info_only:
        print("\n[INFO] --info-only mode, skipping inference tests.")
        return

    if args.image and os.path.exists(args.image):
        image = cv2.imread(args.image)
        if image is None:
            print(f"[ERROR] Cannot read image: {args.image}")
            sys.exit(1)
        print(f"\n[INFO] Using image: {args.image} ({image.shape[1]}x{image.shape[0]})")
    else:
        image = create_synthetic_image()
        print("[INFO] Using synthetic test image (1280x720)")

    # --- Step 2: Raw tensor inference ---
    test_raw_infer(validator)

    # --- Step 3: End-to-end ---
    test_run_e2e(validator, image, args.preproc, args.postproc,
                 args.postproc_config)

    # --- Step 4: Benchmark ---
    if not args.no_benchmark:
        test_benchmark(validator, image, args.preproc, args.postproc,
                       args.postproc_config, args.warmup, args.runs)

    print("\n" + "=" * 60)
    print("  All validation steps completed.")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: modules/unittest/inference/test_model_validator.cpp

#include <cstdlib>
#include <string>
#include <algorithm>

#include "base.hpp"
#include "model_validator.hpp"
#include "cnstream_frame_va.hpp"

using namespace cnstream;


static std::string model_path = "./model/20260625/yolov8s_tracing_static_b1_pre.engine";
static std::string device_type = "cuda";
static std::string image_path = "./image.png";

// Test: DataFrame::SetImage functionality
TEST(ModelValidator, DataFrameSetImage) {
  DataFramePtr frame = std::make_shared<DataFrame>();

  cv::Mat img(480, 640, CV_8UC3, cv::Scalar(1, 2, 3));
  frame->SetImage(img);

  EXPECT_EQ(frame->GetWidth(), 640);
  EXPECT_EQ(frame->GetHeight(), 480);
  EXPECT_TRUE(frame->HasImage());

  cv::Mat recovered = frame->GetImage();
  EXPECT_FALSE(recovered.empty());
  EXPECT_EQ(recovered.cols, 640);
  EXPECT_EQ(recovered.rows, 480);
}

TEST(ModelValidator, LoadNonExistentModel) {
  ModelValidator v(model_path, device_type, 0);
  bool ok = v.Load();
  ASSERT_TRUE(ok);
  ASSERT_TRUE(v.IsLoaded());
}


/* ============================================================ *
 *  Tests that REQUIRE a real model — skipped if env var not set.
 *
 *  Set environment variable to enable:
 *    VSTREAM_TEST_MODEL_PATH=/path/to/yolov8s.engine
 *    VSTREAM_TEST_DEVICE=cuda            (optional, default: cuda)
 *    VSTREAM_TEST_POSTPROC_CONFIG=/path/to/yolo_coco.json  (optional)
 *    VSTREAM_TEST_IMAGE=/path/to/test.jpg (optional, default: synthetic)
 * ============================================================ */

static std::string GetEnv(const char* name, const std::string& default_val = "") {
  const char* val = std::getenv(name);
  return val ? std::string(val) : default_val;
}

static bool HasTestModel() {
  return !GetEnv("VSTREAM_TEST_MODEL_PATH").empty();
}

// Test: Load real model and inspect info
TEST(ModelValidator, RealModelLoadAndGetInfo) {
  if (!HasTestModel()) GTEST_SKIP() << "Set VSTREAM_TEST_MODEL_PATH to run this test";

  std::string model_path = GetEnv("VSTREAM_TEST_MODEL_PATH");
  std::string device = GetEnv("VSTREAM_TEST_DEVICE", "cuda");

  ModelValidator v(model_path, device, 0);
  ASSERT_TRUE(v.Load()) << "Failed to load model: " << model_path;
  EXPECT_TRUE(v.IsLoaded());

  ModelInfo info = v.GetModelInfo();
  EXPECT_GT(info.inputs.size(), 0u);
  EXPECT_GT(info.outputs.size(), 0u);
  EXPECT_GT(info.batch_size, 0);
  EXPECT_GT(info.width, 0);
  EXPECT_GT(info.height, 0);
  EXPECT_EQ(info.channel, 3);

  // Print model info for debugging
  printf("  Model: %s\n", info.model_path.c_str());
  printf("  Device: %s, id=%d\n", info.device_type.c_str(), info.device_id);
  printf("  Batch: %d, Input: %dx%dx%d\n", info.batch_size, info.width, info.height, info.channel);
  for (size_t i = 0; i < info.inputs.size(); ++i) {
    printf("  Input[%zu]: name=%s, shape=[%d,%d,%d,%d], dtype=%s\n",
           i, info.inputs[i].name.c_str(),
           info.inputs[i].shape[0], info.inputs[i].shape[1],
           info.inputs[i].shape[2], info.inputs[i].shape[3],
           info.inputs[i].dtype.c_str());
  }
  for (size_t i = 0; i < info.outputs.size(); ++i) {
    printf("  Output[%zu]: name=%s, shape=[%d,%d,%d,%d], dtype=%s\n",
           i, info.outputs[i].name.c_str(),
           info.outputs[i].shape[0], info.outputs[i].shape[1],
           info.outputs[i].shape[2], info.outputs[i].shape[3],
           info.outputs[i].dtype.c_str());
  }
}

/**
 * @brief 采用随机数据进行模型推理（不包含前处理和后处理）
 */
TEST(ModelValidator, RealModelRawInfer) {
  if (!HasTestModel()) GTEST_SKIP() << "Set VSTREAM_TEST_MODEL_PATH to run this test";

  std::string model_path = GetEnv("VSTREAM_TEST_MODEL_PATH", model_path);
  std::string device = GetEnv("VSTREAM_TEST_DEVICE", "cuda");

  ModelValidator v(model_path, device, 0);
  ASSERT_TRUE(v.Load());

  ModelInfo info = v.GetModelInfo();

  // Build random inputs matching tensor shapes
  std::vector<std::vector<float>> inputs;
  for (size_t i = 0; i < info.inputs.size(); ++i) {
    size_t count = 1;
    for (int d : info.inputs[i].shape) count *= d;
    std::vector<float> data(count);
    for (size_t j = 0; j < count; ++j) {
      data[j] = static_cast<float>(rand()) / RAND_MAX;
    }
    inputs.push_back(std::move(data));
  }

  auto outputs = v.Infer(inputs);
  EXPECT_EQ(outputs.size(), info.outputs.size());

  // Check output sizes
  for (size_t i = 0; i < outputs.size(); ++i) {
    size_t expected = 1;
    for (int d : info.outputs[i].shape) expected *= d;
    EXPECT_EQ(outputs[i].size(), expected);

    for (float val : outputs[i]) {
      EXPECT_FALSE(std::isnan(val)) << "NaN in output[" << i << "]";
      EXPECT_FALSE(std::isinf(val)) << "Inf in output[" << i << "]";
    }

    if (!outputs[i].empty()) {
      float min_v = *std::min_element(outputs[i].begin(), outputs[i].end());
      float max_v = *std::max_element(outputs[i].begin(), outputs[i].end());
      double sum = std::accumulate(outputs[i].begin(), outputs[i].end(), 0.0);
      double mean = sum / outputs[i].size();
      printf("  Output[%zu]: min=%.4f, max=%.4f, mean=%.4f, size=%zu\n",
             i, min_v, max_v, mean, outputs[i].size());
    }
  }
}


TEST(ModelValidator, RealModelRunE2E) {
  if (!HasTestModel()) GTEST_SKIP() << "Set VSTREAM_TEST_MODEL_PATH to run this test";

  std::string model_path = GetEnv("VSTREAM_TEST_MODEL_PATH", model_path);
  std::string device = GetEnv("VSTREAM_TEST_DEVICE", "cuda");
  std::string postproc_config = GetEnv("VSTREAM_TEST_POSTPROC_CONFIG");
  std::string image_path = GetEnv("VSTREAM_TEST_IMAGE", image_path);

  ModelValidator v(model_path, device, 0);
  ASSERT_TRUE(v.Load());

  // Prepare test image
  cv::Mat img;
  if (!image_path.empty()) {
    img = cv::imread(image_path);
    ASSERT_FALSE(img.empty()) << "Failed to read image: " << image_path;
  } else {
    // Synthetic image: 1280x720 BGR
    img = cv::Mat(720, 1280, CV_8UC3, cv::Scalar(50, 100, 150));
    cv::rectangle(img, cv::Rect(100, 100, 200, 200), cv::Scalar(0, 255, 0), 3);
  }

  // Prepare postproc params
  std::map<std::string, std::string> postproc_params;
  if (!postproc_config.empty()) {
    std::string key_config_file = "config_file";
    postproc_params[key_config_file] = postproc_config;
  }

  E2EResult r = v.RunE2E(img, "Pre_YOLO_CPU_v2", "Post_YOLOv8_CPU_v2", {}, postproc_params);

  if (!r.error.empty()) {
    printf("  E2E error: %s\n", r.error.c_str());
    if (postproc_config.empty()) {
      printf("  (postproc config not provided, skipping detection check)\n");
      return;
    }
  }

  printf("  E2E: %zu detections, %.2f ms\n", r.detections.size(), r.latency_ms);
  for (size_t i = 0; i < r.detections.size() && i < 10; ++i) {
    const auto& d = r.detections[i];
    printf("    det[%zu]: class=%d, score=%.4f, bbox=[%.3f, %.3f, %.3f, %.3f]\n",
           i, d.class_id, d.score, d.x, d.y, d.w, d.h);
  }

  EXPECT_GE(r.latency_ms, 0.0);
}

// Test: Benchmark performance
TEST(ModelValidator, RealModelBenchmark) {
  if (!HasTestModel()) GTEST_SKIP() << "Set VSTREAM_TEST_MODEL_PATH to run this test";

  std::string model_path = GetEnv("VSTREAM_TEST_MODEL_PATH", model_path);
  std::string device = GetEnv("VSTREAM_TEST_DEVICE", "cuda");
  std::string postproc_config = GetEnv("VSTREAM_TEST_POSTPROC_CONFIG");

  ModelValidator v(model_path, device, 0);
  ASSERT_TRUE(v.Load());

  cv::Mat img(720, 1280, CV_8UC3, cv::Scalar(50, 100, 150));

  std::map<std::string, std::string> postproc_params;
  if (!postproc_config.empty()) {
    std::string key_config_file = "config_file";
    postproc_params[key_config_file] = postproc_config;
  }

  auto results = v.Benchmark(img, "Pre_YOLO_CPU_v2", "Post_YOLOv8_CPU_v2",
                              {}, postproc_params,
                              3, 10, {1});

  if (!results.empty()) {
    const auto& r = results[0];
    printf("  Benchmark bs=%d: avg=%.2fms, min=%.2fms, max=%.2fms, p99=%.2fms, fps=%.1f, errors=%d\n",
           r.batch_size, r.avg_ms, r.min_ms, r.max_ms, r.p99_ms, r.fps, r.error_count);
    EXPECT_GT(r.avg_ms, 0.0);
    EXPECT_GE(r.max_ms, r.min_ms);
  }
}


